// docusaurus-theme-openapi-docs/lib/theme/translationIds.js
"use strict";
/* ============================================================================
 * Copyright (c) Palo Alto Networks
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OPENAPI_SECURITY_SCHEMES =
  exports.OPENAPI_PARAMS_DETAILS =
  exports.OPENAPI_SCHEMA_ITEM =
  exports.OPENAPI_SCHEMA =
  exports.OPENAPI_STATUS_CODES =
  exports.OPENAPI_BODY =
  exports.OPENAPI_RESPONSE_EXAMPLES =
  exports.OPENAPI_AUTH =
  exports.OPENAPI_FORM =
  exports.OPENAPI_FORM_FILE_UPLOAD =
  exports.OPENAPI_PARAM_OPTIONS =
  exports.OPENAPI_SERVER =
  exports.OPENAPI_REQUEST =
  exports.OPENAPI_TABS =
  exports.OPENAPI_RESPONSE =
    void 0;
exports.OPENAPI_RESPONSE = {
  TITLE: "theme.openapi.response.title",
  CLEAR: "theme.openapi.response.clear",
  PLACEHOLDER: "theme.openapi.response.placeholder",
  HEADERS_TAB: "theme.openapi.response.headersTab",
};
exports.OPENAPI_TABS = {
  RESPONSES_LABEL: "theme.openapi.tabs.responses.label",
};
exports.OPENAPI_REQUEST = {
  BODY_TITLE: "theme.openapi.request.body.title",
  ACCEPT_TITLE: "theme.openapi.request.accept.title",
  SEND_BUTTON: "theme.openapi.request.sendButton",
  REQUIRED_LABEL: "theme.openapi.request.requiredLabel",
  REQUEST_TITLE: "theme.openapi.request.title",
  COLLAPSE_ALL: "theme.openapi.request.collapseAll",
  EXPAND_ALL: "theme.openapi.request.expandAll",
  BASE_URL_TITLE: "theme.openapi.request.baseUrl.title",
  AUTH_TITLE: "theme.openapi.request.auth.title",
  PARAMETERS_TITLE: "theme.openapi.request.parameters.title",
  FETCHING_MESSAGE: "theme.openapi.request.fetchingMessage",
  CONNECTION_FAILED: "theme.openapi.request.connectionFailed",
  ERROR_TIMEOUT: "theme.openapi.request.error.timeout",
  ERROR_NETWORK: "theme.openapi.request.error.network",
  ERROR_CORS: "theme.openapi.request.error.cors",
  ERROR_UNKNOWN: "theme.openapi.request.error.unknown",
};
exports.OPENAPI_SERVER = {
  EDIT_BUTTON: "theme.openapi.server.editButton",
  HIDE_BUTTON: "theme.openapi.server.hideButton",
};
exports.OPENAPI_PARAM_OPTIONS = {
  SHOW_OPTIONAL: "theme.openapi.paramOptions.showOptional",
  HIDE_OPTIONAL: "theme.openapi.paramOptions.hideOptional",
};
exports.OPENAPI_FORM_FILE_UPLOAD = {
  CLEAR_BUTTON: "theme.openapi.formFileUpload.clearButton",
};
exports.OPENAPI_FORM = {
  FIELD_REQUIRED: "theme.openapi.form.fieldRequired",
};
exports.OPENAPI_AUTH = {
  BEARER_TOKEN: "theme.openapi.auth.bearerToken",
  USERNAME: "theme.openapi.auth.username",
  PASSWORD: "theme.openapi.auth.password",
  SECURITY_SCHEME: "theme.openapi.auth.securityScheme",
};
exports.OPENAPI_RESPONSE_EXAMPLES = {
  EXAMPLE: "theme.openapi.responseExamples.example",
  AUTO_EXAMPLE: "theme.openapi.responseExamples.autoExample",
};
exports.OPENAPI_BODY = {
  EXAMPLE_FROM_SCHEMA: "theme.openapi.body.exampleFromSchema",
};
exports.OPENAPI_STATUS_CODES = {
  RESPONSE_HEADERS: "theme.openapi.statusCodes.responseHeaders",
  SCHEMA_TITLE: "theme.openapi.statusCodes.schemaTitle",
};
exports.OPENAPI_SCHEMA = {
  NO_SCHEMA: "theme.openapi.schema.noSchema",
};
exports.OPENAPI_SCHEMA_ITEM = {
  REQUIRED: "theme.openapi.schemaItem.required",
  DEPRECATED: "theme.openapi.schemaItem.deprecated",
  NULLABLE: "theme.openapi.schemaItem.nullable",
  DEFAULT_VALUE: "theme.openapi.schemaItem.defaultValue",
  EXAMPLE: "theme.openapi.schemaItem.example",
  EXAMPLES: "theme.openapi.schemaItem.examples",
  DESCRIPTION: "theme.openapi.schemaItem.description",
  CONSTANT_VALUE: "theme.openapi.schemaItem.constantValue",
  ENUM_VALUE: "theme.openapi.schemaItem.enumValue",
  ENUM_DESCRIPTION: "theme.openapi.schemaItem.enumDescription",
  POSSIBLE_VALUES: "theme.openapi.schemaItem.possibleValues",
  ONE_OF: "theme.openapi.schemaItem.oneOf",
  ANY_OF: "theme.openapi.schemaItem.anyOf",
};
exports.OPENAPI_PARAMS_DETAILS = {
  PARAMETERS_TITLE: "theme.openapi.paramsDetails.parametersTitle",
};
exports.OPENAPI_SECURITY_SCHEMES = {
  NAME: "theme.openapi.securitySchemes.name",
  TYPE: "theme.openapi.securitySchemes.type",
  SCOPES: "theme.openapi.securitySchemes.scopes",
  IN: "theme.openapi.securitySchemes.in",
  FLOWS: "theme.openapi.securitySchemes.flows",
  DESCRIPTION: "theme.openapi.securitySchemes.description",
  SCHEME: "theme.openapi.securitySchemes.scheme",
  BEARER_FORMAT: "theme.openapi.securitySchemes.bearerFormat",
  OPEN_ID_CONNECT_URL: "theme.openapi.securitySchemes.openIdConnectUrl",
};
