// docusaurus-theme-openapi-docs/lib/theme/ApiExplorer/Body/FormBodyItem/index.js
"use strict";
/* ============================================================================
 * Copyright (c) Palo Alto Networks
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */
var __importDefault =
  (this && this.__importDefault) ||
  function (mod) {
    return mod && mod.__esModule ? mod : { default: mod };
  };
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = FormBodyItem;
const react_1 = __importDefault(require("react"));
const FormFileUpload_1 = __importDefault(
  require("@theme/ApiExplorer/FormFileUpload")
);
const FormSelect_1 = __importDefault(require("@theme/ApiExplorer/FormSelect"));
const FormTextInput_1 = __importDefault(
  require("@theme/ApiExplorer/FormTextInput")
);
const LiveEditor_1 = __importDefault(require("@theme/ApiExplorer/LiveEditor"));
const hooks_1 = require("@theme/ApiItem/hooks");
const FileArrayFormBodyItem_1 = __importDefault(
  require("../FileArrayFormBodyItem")
);
const slice_1 = require("../slice");
function FormBodyItem({ schemaObject, id, schema }) {
  const dispatch = (0, hooks_1.useTypedDispatch)();
  if (
    schemaObject.type === "array" &&
    schemaObject.items?.format === "binary"
  ) {
    return react_1.default.createElement(FileArrayFormBodyItem_1.default, {
      id: id,
      description: schemaObject.description,
    });
  }
  if (schemaObject.format === "binary") {
    return react_1.default.createElement(FormFileUpload_1.default, {
      placeholder: schemaObject.description || id,
      onChange: (file) => {
        if (file === undefined) {
          dispatch((0, slice_1.clearFormBodyKey)(id));
          return;
        }
        dispatch(
          (0, slice_1.setFileFormBody)({
            key: id,
            value: {
              src: `/path/to/${file.name}`,
              content: file,
            },
          })
        );
      },
    });
  }
  if (
    schemaObject.type === "object" &&
    (schemaObject.example || schemaObject.examples)
  ) {
    const objectExample = JSON.stringify(
      schemaObject.example ?? schemaObject.examples[0],
      null,
      2
    );
    return react_1.default.createElement(
      LiveEditor_1.default,
      {
        action: (code) =>
          dispatch((0, slice_1.setStringFormBody)({ key: id, value: code })),
      },
      objectExample
    );
  }
  if (
    schemaObject.enum &&
    schemaObject.enum.every((value) => typeof value === "string")
  ) {
    return react_1.default.createElement(FormSelect_1.default, {
      options: ["---", ...schemaObject.enum],
      onChange: (e) => {
        const val = e.target.value;
        if (val === "---") {
          dispatch((0, slice_1.clearFormBodyKey)(id));
        } else {
          dispatch(
            (0, slice_1.setStringFormBody)({
              key: id,
              value: val,
            })
          );
        }
      },
    });
  }
  // TODO: support all the other types.
  return react_1.default.createElement(FormTextInput_1.default, {
    paramName: id,
    isRequired: Array.isArray(schema.required) && schema.required.includes(id),
    placeholder: schemaObject.description || id,
    onChange: (e) => {
      dispatch(
        (0, slice_1.setStringFormBody)({ key: id, value: e.target.value })
      );
    },
  });
}
