// docusaurus-theme-openapi-docs/lib/theme/ApiExplorer/SchemaSelection/slice.js
"use strict";
/* ============================================================================
 * Copyright (c) Palo Alto Networks
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.clearSchemaSelections =
  exports.setSchemaSelection =
  exports.slice =
    void 0;
const toolkit_1 = require("@reduxjs/toolkit");
const initialState = {
  selections: {},
};
exports.slice = (0, toolkit_1.createSlice)({
  name: "schemaSelection",
  initialState,
  reducers: {
    /**
     * Set the selected index for a specific schema path
     */
    setSchemaSelection: (state, action) => {
      state.selections[action.payload.path] = action.payload.index;
    },
    /**
     * Clear all schema selections (useful when navigating to a new API endpoint)
     */
    clearSchemaSelections: (state) => {
      state.selections = {};
    },
  },
});
((_a = exports.slice.actions),
  (exports.setSchemaSelection = _a.setSchemaSelection),
  (exports.clearSchemaSelections = _a.clearSchemaSelections));
exports.default = exports.slice.reducer;
