// docusaurus-theme-openapi-docs/lib/theme/ApiExplorer/SecuritySchemes/index.js
"use strict";
/* ============================================================================
 * Copyright (c) Palo Alto Networks
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */
var __importDefault =
  (this && this.__importDefault) ||
  function (mod) {
    return mod && mod.__esModule ? mod : { default: mod };
  };
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const Link_1 = __importDefault(require("@docusaurus/Link"));
const Translate_1 = require("@docusaurus/Translate");
const hooks_1 = require("@theme/ApiItem/hooks");
const translationIds_1 = require("@theme/translationIds");
function SecuritySchemes(props) {
  const options = (0, hooks_1.useTypedSelector)((state) => state.auth.options);
  const selected = (0, hooks_1.useTypedSelector)(
    (state) => state.auth.selected
  );
  const infoAuthPath = `/${props.infoPath}#authentication`;
  if (selected === undefined) return null;
  if (options[selected]?.[0]?.type === undefined) {
    return null;
  }
  const selectedAuth = options[selected];
  const keyTranslations = {
    description: {
      id: translationIds_1.OPENAPI_SECURITY_SCHEMES.DESCRIPTION,
      message: "description:",
    },
    scheme: {
      id: translationIds_1.OPENAPI_SECURITY_SCHEMES.SCHEME,
      message: "scheme:",
    },
    bearerFormat: {
      id: translationIds_1.OPENAPI_SECURITY_SCHEMES.BEARER_FORMAT,
      message: "bearerFormat:",
    },
    openIdConnectUrl: {
      id: translationIds_1.OPENAPI_SECURITY_SCHEMES.OPEN_ID_CONNECT_URL,
      message: "openIdConnectUrl:",
    },
  };
  const renderRest = (rest) =>
    Object.keys(rest).map((k) => {
      const translation = keyTranslations[k];
      const label = translation
        ? (0, Translate_1.translate)({
            id: translation.id,
            message: translation.message,
          })
        : `${k}:`;
      return react_1.default.createElement(
        "span",
        { key: k },
        react_1.default.createElement("strong", null, label, " "),
        typeof rest[k] === "object"
          ? JSON.stringify(rest[k], null, 2)
          : String(rest[k])
      );
    });
  return react_1.default.createElement(
    "details",
    { className: "openapi-security__details", open: false },
    react_1.default.createElement(
      "summary",
      { className: "openapi-security__summary-container" },
      react_1.default.createElement(
        "h4",
        { className: "openapi-security__summary-header" },
        "Authorization: ",
        selectedAuth[0].name ?? selectedAuth[0].type
      )
    ),
    selectedAuth.map((auth) => {
      const isHttp = auth.type === "http";
      const isApiKey = auth.type === "apiKey";
      const isOauth2 = auth.type === "oauth2";
      const isOpenId = auth.type === "openIdConnect";
      if (isHttp) {
        if (auth.scheme === "bearer") {
          const { name, key, type, scopes, ...rest } = auth;
          return react_1.default.createElement(
            react_1.default.Fragment,
            { key: auth.key },
            react_1.default.createElement(
              "pre",
              {
                style: {
                  display: "flex",
                  flexDirection: "column",
                  background: "var(--openapi-card-background-color)",
                },
              },
              react_1.default.createElement(
                "span",
                null,
                react_1.default.createElement(
                  "strong",
                  null,
                  (0, Translate_1.translate)({
                    id: translationIds_1.OPENAPI_SECURITY_SCHEMES.NAME,
                    message: "name:",
                  })
                ),
                " ",
                react_1.default.createElement(
                  Link_1.default,
                  { to: infoAuthPath },
                  name ?? key
                )
              ),
              react_1.default.createElement(
                "span",
                null,
                react_1.default.createElement(
                  "strong",
                  null,
                  (0, Translate_1.translate)({
                    id: translationIds_1.OPENAPI_SECURITY_SCHEMES.TYPE,
                    message: "type:",
                  })
                ),
                " ",
                type
              ),
              scopes &&
                scopes.length > 0 &&
                react_1.default.createElement(
                  "span",
                  null,
                  react_1.default.createElement(
                    "strong",
                    null,
                    (0, Translate_1.translate)({
                      id: translationIds_1.OPENAPI_SECURITY_SCHEMES.SCOPES,
                      message: "scopes:",
                    })
                  ),
                  " ",
                  react_1.default.createElement(
                    "code",
                    null,
                    auth.scopes.length > 0 ? auth.scopes.toString() : "[]"
                  )
                ),
              renderRest(rest)
            )
          );
        }
        if (auth.scheme === "basic") {
          const { name, key, type, scopes, ...rest } = auth;
          return react_1.default.createElement(
            react_1.default.Fragment,
            { key: auth.key },
            react_1.default.createElement(
              "pre",
              {
                style: {
                  display: "flex",
                  flexDirection: "column",
                  background: "var(--openapi-card-background-color)",
                },
              },
              react_1.default.createElement(
                "span",
                null,
                react_1.default.createElement(
                  "strong",
                  null,
                  (0, Translate_1.translate)({
                    id: translationIds_1.OPENAPI_SECURITY_SCHEMES.NAME,
                    message: "name:",
                  })
                ),
                " ",
                react_1.default.createElement(
                  Link_1.default,
                  { to: infoAuthPath },
                  name ?? key
                )
              ),
              react_1.default.createElement(
                "span",
                null,
                react_1.default.createElement(
                  "strong",
                  null,
                  (0, Translate_1.translate)({
                    id: translationIds_1.OPENAPI_SECURITY_SCHEMES.TYPE,
                    message: "type:",
                  })
                ),
                " ",
                type
              ),
              scopes &&
                scopes.length > 0 &&
                react_1.default.createElement(
                  "span",
                  null,
                  react_1.default.createElement(
                    "strong",
                    null,
                    (0, Translate_1.translate)({
                      id: translationIds_1.OPENAPI_SECURITY_SCHEMES.SCOPES,
                      message: "scopes:",
                    })
                  ),
                  " ",
                  react_1.default.createElement(
                    "code",
                    null,
                    auth.scopes.length > 0 ? auth.scopes.toString() : "[]"
                  )
                ),
              renderRest(rest)
            )
          );
        }
        return react_1.default.createElement(
          react_1.default.Fragment,
          { key: auth.key },
          react_1.default.createElement(
            "pre",
            {
              style: {
                display: "flex",
                flexDirection: "column",
                background: "var(--openapi-card-background-color)",
              },
            },
            react_1.default.createElement(
              "span",
              null,
              react_1.default.createElement(
                "strong",
                null,
                (0, Translate_1.translate)({
                  id: translationIds_1.OPENAPI_SECURITY_SCHEMES.NAME,
                  message: "name:",
                })
              ),
              " ",
              react_1.default.createElement(
                Link_1.default,
                { to: infoAuthPath },
                auth.name ?? auth.key
              )
            ),
            react_1.default.createElement(
              "span",
              null,
              react_1.default.createElement(
                "strong",
                null,
                (0, Translate_1.translate)({
                  id: translationIds_1.OPENAPI_SECURITY_SCHEMES.TYPE,
                  message: "type:",
                })
              ),
              " ",
              auth.type
            ),
            react_1.default.createElement(
              "span",
              null,
              react_1.default.createElement(
                "strong",
                null,
                (0, Translate_1.translate)({
                  id: translationIds_1.OPENAPI_SECURITY_SCHEMES.IN,
                  message: "in:",
                })
              ),
              " ",
              auth.in
            )
          )
        );
      }
      if (isApiKey) {
        const { name, key, type, scopes, ...rest } = auth;
        return react_1.default.createElement(
          react_1.default.Fragment,
          { key: auth.key },
          react_1.default.createElement(
            "pre",
            {
              style: {
                display: "flex",
                flexDirection: "column",
                background: "var(--openapi-card-background-color)",
              },
            },
            react_1.default.createElement(
              "span",
              null,
              react_1.default.createElement(
                "strong",
                null,
                (0, Translate_1.translate)({
                  id: translationIds_1.OPENAPI_SECURITY_SCHEMES.NAME,
                  message: "name:",
                })
              ),
              " ",
              react_1.default.createElement(
                Link_1.default,
                { to: infoAuthPath },
                name ?? key
              )
            ),
            react_1.default.createElement(
              "span",
              null,
              react_1.default.createElement(
                "strong",
                null,
                (0, Translate_1.translate)({
                  id: translationIds_1.OPENAPI_SECURITY_SCHEMES.TYPE,
                  message: "type:",
                })
              ),
              " ",
              type
            ),
            scopes &&
              scopes.length > 0 &&
              react_1.default.createElement(
                "span",
                null,
                react_1.default.createElement(
                  "strong",
                  null,
                  (0, Translate_1.translate)({
                    id: translationIds_1.OPENAPI_SECURITY_SCHEMES.SCOPES,
                    message: "scopes:",
                  })
                ),
                " ",
                react_1.default.createElement(
                  "code",
                  null,
                  auth.scopes.length > 0 ? auth.scopes.toString() : "[]"
                )
              ),
            renderRest(rest)
          )
        );
      }
      if (isOauth2) {
        const { name, key, type, scopes, flows, ...rest } = auth;
        return react_1.default.createElement(
          react_1.default.Fragment,
          { key: selected },
          react_1.default.createElement(
            "pre",
            {
              style: {
                display: "flex",
                flexDirection: "column",
                background: "var(--openapi-card-background-color)",
              },
            },
            react_1.default.createElement(
              "span",
              null,
              react_1.default.createElement(
                "strong",
                null,
                (0, Translate_1.translate)({
                  id: translationIds_1.OPENAPI_SECURITY_SCHEMES.NAME,
                  message: "name:",
                })
              ),
              " ",
              react_1.default.createElement(
                Link_1.default,
                { to: infoAuthPath },
                name ?? key
              )
            ),
            react_1.default.createElement(
              "span",
              null,
              react_1.default.createElement(
                "strong",
                null,
                (0, Translate_1.translate)({
                  id: translationIds_1.OPENAPI_SECURITY_SCHEMES.TYPE,
                  message: "type:",
                })
              ),
              " ",
              type
            ),
            scopes &&
              scopes.length > 0 &&
              react_1.default.createElement(
                "span",
                null,
                react_1.default.createElement(
                  "strong",
                  null,
                  (0, Translate_1.translate)({
                    id: translationIds_1.OPENAPI_SECURITY_SCHEMES.SCOPES,
                    message: "scopes:",
                  })
                ),
                " ",
                react_1.default.createElement(
                  "code",
                  null,
                  auth.scopes.length > 0 ? auth.scopes.toString() : "[]"
                )
              ),
            renderRest(rest),
            flows &&
              react_1.default.createElement(
                "span",
                null,
                react_1.default.createElement(
                  "code",
                  null,
                  react_1.default.createElement(
                    "strong",
                    null,
                    (0, Translate_1.translate)({
                      id: translationIds_1.OPENAPI_SECURITY_SCHEMES.FLOWS,
                      message: "flows:",
                    })
                  ),
                  " ",
                  JSON.stringify(flows, null, 2)
                )
              )
          )
        );
      }
      if (isOpenId) {
        const { name, key, scopes, type, ...rest } = auth;
        return react_1.default.createElement(
          react_1.default.Fragment,
          { key: auth.key },
          react_1.default.createElement(
            "pre",
            {
              style: {
                display: "flex",
                flexDirection: "column",
                background: "var(--openapi-card-background-color)",
              },
            },
            react_1.default.createElement(
              "span",
              null,
              react_1.default.createElement(
                "strong",
                null,
                (0, Translate_1.translate)({
                  id: translationIds_1.OPENAPI_SECURITY_SCHEMES.NAME,
                  message: "name:",
                })
              ),
              " ",
              react_1.default.createElement(
                Link_1.default,
                { to: infoAuthPath },
                name ?? key
              )
            ),
            react_1.default.createElement(
              "span",
              null,
              react_1.default.createElement(
                "strong",
                null,
                (0, Translate_1.translate)({
                  id: translationIds_1.OPENAPI_SECURITY_SCHEMES.TYPE,
                  message: "type:",
                })
              ),
              " ",
              type
            ),
            scopes &&
              scopes.length > 0 &&
              react_1.default.createElement(
                "span",
                null,
                react_1.default.createElement(
                  "strong",
                  null,
                  (0, Translate_1.translate)({
                    id: translationIds_1.OPENAPI_SECURITY_SCHEMES.SCOPES,
                    message: "scopes:",
                  })
                ),
                " ",
                react_1.default.createElement(
                  "code",
                  null,
                  auth.scopes.length > 0 ? auth.scopes.toString() : "[]"
                )
              ),
            renderRest(rest)
          )
        );
      }
      return undefined;
    })
  );
}
exports.default = SecuritySchemes;
